crypto/merklearray/merkle.go

```diff
@@ -29,53 +30,62 @@ type Tree struct {
 	_struct struct{} `codec:",omitempty,omitemptyarray"`
 
 	// Level 0 is the leaves.
-	Levels []Layer `codec:"lvls,allocbound=-"`
+	Levels []Layer            `codec:"lvls,allocbound=-"`
+	Hash   crypto.HashFactory `codec:"hsh"`
 }
 
 func (tree *Tree) topLayer() Layer {
 	return tree.Levels[len(tree.Levels)-1]
 }
 
-func buildWorker(ws *workerState, array Array, leaves Layer, errs chan error) {
+type errorChannel chan error
+
+func (ch errorChannel) nonBlockingSend(e error) {
+	select {
+	case ch <- e:
+	default:
+	}
+}
+
+func buildWorker(ws *workerState, array Array, leaves Layer, h crypto.HashFactory, errs errorChannel) {
+	defer ws.done()
```

Comment:
please move this after the start

Reply:
i think this should be the first thing in the function to state to everyone, that once this function is finished, it removes work from the workerState.